File: src/test/java/com/huawei/wallet/hms/eventticket/EventTicketModelTest.java
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.huawei.wallet.hms.eventticket;

import com.huawei.wallet.hms.ServerApiService;
import com.huawei.wallet.hms.ServerApiServiceImpl;
import com.huawei.wallet.util.ConfigUtil;
import com.huawei.wallet.util.HwWalletObjectUtil;

import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONObject;

import org.junit.Test;

/**
 * Event ticket model tests.
 *
 * @since 2019-12-12
 */
public class EventTicketModelTest {
    private final ServerApiService serverApiService = new ServerApiServiceImpl();

    /**
     * Create a new event ticket model.
     * Each event ticket model indicates a style of event ticket passes.
     * POST http://XXX/hmspass/v1/eventticket/model
     */
    @Test
    public void createEventTicketModel() {
        System.out.println("createEventTicketModel begin.");

        // Read an event ticket model from a JSON file.
        JSONObject model = JSONObject.parseObject(ConfigUtil.readFile("EventTicketModel.json"));

        // Validate parameters.
        HwWalletObjectUtil.validateModel(model);

        // Post the new event ticket model to HMS wallet server.
        String urlSegment = "eventticket/model";
        JSONObject responseModel = serverApiService.postToWalletServer(urlSegment, JSONObject.toJSONString(model));
        System.out.println("Posted event ticket model: " + JSONObject.toJSONString(responseModel));
    }

    /**
     * Get an event ticket model by its model ID.
     * Run the "createEventTicketModel" test before running this test.
     * GET http://xxx/hmspass/v1/eventticket/model/{modelId}
     */
    @Test
    public void getEventTicketModel() {
        System.out.println("getEventTicketModel begin.");

        // ID of the event ticket model you want to get.
        String modelId = "EventTicketModelTest";

        // Get the event ticket model.
        String urlSegment = "eventticket/model/";
        JSONObject responseModel = serverApiService.getHwWalletObjectById(urlSegment, modelId);
        System.out.println("Corresponding event ticket model: " + JSONObject.toJSONString(responseModel));
    }

    /**
     * Get event ticket models belonging to a specific appId.
     * Run the "createEventTicketModel" test before running this test.
     * GET http://xxx/hmspass/v1/eventticket/model?session=XXX&pageSize=XXX
     */
    @Test
    public void getEventTicketModelList() {
        System.out.println("getEventTicketModelList begin.");

        // Get a list of event ticket models.
        String urlSegment = "eventticket/model";

        JSONArray models = serverApiService.getModels(urlSegment, 5);
        System.out.println("Total models count: " + models.size());
        System.out.println("Models list: " + models.toJSONString());
    }

    /**
     * Overwrite an event ticket model.
     * Run the "createEventTicketModel" test before running this test.
     * PUT http://xxx/hmspass/v1/eventticket/model/{modelId}
     */
    @Test
    public void fullUpdateEventTicketModel() {
        System.out.println("fullUpdateEventTicketModel begin.");

        // Read a HwWalletObject from a JSON file. This HwWalletObject will overwrite the corresponding model.
        JSONObject model = JSONObject.parseObject(ConfigUtil.readFile("FullUpdateEventTicketModel.json"));

        // Validate parameters.
        HwWalletObjectUtil.validateModel(model);

        // Update the event ticket model.
        String urlSegment = "eventticket/model/";
        JSONObject responseModel = serverApiService.fullUpdateHwWalletObject(urlSegment,
            model.getString("passStyleIdentifier"), JSONObject.toJSONString(model));
        System.out.println("Updated event ticket model: " + JSONObject.toJSONString(responseModel));
    }

    /**
     * Update an event ticket model.
     * Run the "createEventTicketModel" test before running this test.
     * PATCH http://xxx/hmspass/v1/eventticket/model/{modelId}
     */
    @Test
    public void partialUpdateEventTicketModel() {
        System.out.println("partialUpdateEventTicketModel begin.");

        // ID of the event ticket model you want to update.
        String modelId = "eventTicketModelTest";

        // Read a HwWalletObject from a JSON file. This HwWalletObject will merge with the corresponding model.
        String modelStr = ConfigUtil.readFile("PartialUpdateEventTicketModel.json");

        // Update the event ticket model.
        String urlSegment = "eventticket/model/";
        JSONObject responseModel = serverApiService.partialUpdateHwWalletObject(urlSegment, modelId, modelStr);
        System.out.println("Updated event ticket model: " + JSONObject.toJSONString(responseModel));
    }

    /**
     * Add messages to an event ticket model.
     * Run the "createEventTicketModel" test before running this test.
     * POST http://xxx/hmspass/v1/eventticket/model/{modelId}/addMessage
     */
    @Test
    public void addMessageToEventTicketModel() {
        System.out.println("addMessageToEventTicketModel begin.");

        // ID of the event ticket model you want to update.
        String modelId = "eventTicketModelTest";

        // Create a list of messages you want to add to a model. Each message contains key, value, and label.
        // The list should not contain multiple messages with the same key. You can update an existing message by adding
        // a new message with the same key. One model contains at most 10 messages. If a model already have 10 messages
        // and you keep adding new messages, the oldest messages will be removed. You should not add more than 10
        // messages at a time.

        // Read messages from a JSON file.
        String messagesStr = ConfigUtil.readFile("Messages.json");

        // Add messages to the event ticket model.
        String urlSegment = "eventticket/model/";
        JSONObject responseModel = serverApiService.addMessageToHwWalletObject(urlSegment, modelId, messagesStr);
        System.out.println("Updated event ticket model: " + JSONObject.toJSONString(responseModel));
    }
}

File: src/demo/eventticket/eventticketmodel_test.go
package eventticket

import (
	"../../../src/api"
	"../../model"
	"fmt"
	"io/ioutil"
	"path/filepath"
	"testing"
)

// Create a new event ticket model.
// Each event ticket model indicates a style of event ticket passes..
// POST http://XXX/hmspass/v1/eventticket/model
func TestCreateEventTicketModel(t *testing.T) {
	t.Log("createEventTicketModel begin.")

	// Read an event ticket model from a JSON file.
	dataPath, _ := filepath.Abs("../data/EventTicketModel.json")
	dataByte, _ := ioutil.ReadFile(dataPath)

	// Post the new event ticket model to HMS wallet server.
	urlSegment := "eventticket/model"
	resp, err := api.PostToWalletServer(urlSegment, dataByte)
	if err != nil {
		t.Logf("Add event ticket model failed. %v", err)
	} else {
		t.Logf("Add event ticket model success, eventticket/model response: %v", resp)
	}
}

// Get an event ticket model by its model ID.
// Run the "createEventTicketModel" test before running this test.
// GET http://xxx/hmspass/v1/eventticket/model/{modelId}
func TestGetEventTicketModel(t *testing.T) {
	t.Log("getEventTicketModel begin.")

	// ID of the event ticket model you want to get.
	modelId := "EventTicketTestModel"

	// Get the event ticket model.
	urlSegment := fmt.Sprintf("eventticket/model/%s", modelId)
	resp, err := api.GetHwWalletObjectById(urlSegment)
	if err != nil {
		t.Logf("Get event ticket model failed. %v", err)
	} else {
		t.Logf("Get event ticket model success, eventticket/model response: %v", resp)
	}
}

// Get event ticket models belonging to a specific appId.
// Run the "createEventTicketModel" test before running this test.
// GET http://xxx/hmspass/v1/eventticket/model?session=XXX&pageSize=XXX
func TestGetEventTicketModelList(t *testing.T) {
	t.Log("getEventTicketModelList begin.")

	// Get a list of event ticket models.
	urlSegment := "eventticket/model"
	api.GetModels(urlSegment, 5)
}

// Overwrite an event ticket model.
// Run the "createEventTicketModel" test before running this test.
// PUT http://xxx/hmspass/v1/eventticket/model/{modelId}
func TestFullUpdateEventTicketModel(t *testing.T) {
	t.Log("fullUpdateEventTicketModel begin.")

	// Read an event ticket model from a JSON file.
	dataPath, _ := filepath.Abs("../data/FullUpdateEventTicketModel.json")
	dataByte, _ := ioutil.ReadFile(dataPath)

	var hwWalletObject model.HwWalletObject
	api.JSONByteToStruct(dataByte, &hwWalletObject)

	// Post the new event ticket model to HMS wallet server.
	urlSegment := fmt.Sprintf("eventticket/model/%s", hwWalletObject.PassStyleIdentifier)
	resp, err := api.FullUpdateHwWalletObject(urlSegment, dataByte)
	if err != nil {
		t.Logf("Full update event ticket model failed. %v", err)
	} else {
		t.Logf("Full update ticket model success, response: %v", resp)
	}
}

// Update an event ticket model.
// Run the "createEventTicketModel" test before running this test.
// PATCH http://xxx/hmspass/v1/eventticket/model/{modelId}
func TestPartialUpdateEventTicketModel(t *testing.T) {
	t.Log("partialUpdateEventTicketModel begin.")

	// ID of the event ticket model you want to update.
	modelId := "EventTicketTestModel"

	// Read a HwWalletObject from a JSON file. This HwWalletObject will merge with the corresponding model.
	dataPath, _ := filepath.Abs("../data/PartialUpdateEventTicketModel.json")
	dataByte, _ := ioutil.ReadFile(dataPath)

	// Post the new event ticket model to HMS wallet server.
	urlSegment := fmt.Sprintf("eventticket/model/%s", modelId)
	resp, err := api.PartialUpdateHwWalletObject(urlSegment, dataByte)
	if err != nil {
		t.Logf("Partial update event ticket model failed. %v", err)
	} else {
		t.Logf("Partial update event ticket model success, response: %v", resp)
	}
}

// Add messages to an event ticket model.
// Run the "createEventTicketModel" test before running this test.
// POST http://xxx/hmspass/v1/eventticket/model/{modelId}/addMessage
func TestAddMessageToEventTicketModel(t *testing.T) {
	// ID of the event ticket model you want to update.
	modelId := "EventTicketTestModel"

	// Read messages from a JSON file.
	dataPath, _ := filepath.Abs("../data/Messages.json")
	dataByte, _ := ioutil.ReadFile(dataPath)

	// Add messages to the event ticket instance.
	urlSegment := fmt.Sprintf("eventticket/model/%s/addMessage", modelId)
	resp, err := api.AddMessageToHwWalletObject(urlSegment, dataByte)
	if err != nil {
		t.Logf("Update message to event ticket model failed. %v", err)
	} else {
		t.Logf("Update message to event ticket model success, response: %v", resp)
	}
}
